Use ip address from CNI output

AlexeyPerevalov opened the pull request:
Before this patch Kubernetes obtained ip address from container's eth
device. But in case when containers work only with loopback device and
with vhostuser (it's not device it's UNIX domain socket file) Kubernetes
can't rely on this approach, it should rely on CNI output. Most of plugins
provide IP address of created port.

This patch also touches unit test, CNI spec declares ip field as mandatory
field of ADD/GET command output. But some users may relay on current behaviour
so code where ip is empty and obtained from container's eth is preserved,
but it's not covered in unit test.

Signed-off-by: Alexey Perevalov <a.perevalov@samsung.com>
Fixes #65756

This is a third version of https://github.com/kubernetes/kubernetes/pull/65759 and https://github.com/kubernetes/kubernetes/pull/68876
I wasn't able to update my current PR.

Current implementation has functional drawback: if we restart kubelet when pod already started,
pod's IP becames none, after that pod will be restarted, and its IP obtained by CNI ADD method
is correct. But restarting of the pod is not desirable behaviour.

Update: after moving to libcni 0.7.0 we don't have drawback described above!

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

Commit:
Use ip address from CNI output

Before this patch Kubernetes obtained ip address from container's eth
device. But in case when containers work only with loopback device and
with vhostuser (it's not device it's UNIX domain socket file) Kubernetes
can't rely on this approach, it should rely on CNI output. Most of plugins
provide IP address of created port.

This patch also touches unit test, CNI spec declares ip field as mandatory
field of ADD/GET command output.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

AlexeyPerevalov commented:
/sig network

AlexeyPerevalov commented:
/assign @dcbw

idealhack commented:
/kind bug
/release-note-none
/ok-to-test

PatrickLang commented:
@feiskyer will this affect Windows?

caseydavenport commented:
> Current implementation has functional drawback: if we restart kubelet when pod already started,
pod's IP becames none, after that pod will be restarted, and its IP obtained by CNI ADD method
is correct. But restarting of the pod is not desirable behaviour.

I think this is really substantial, and we'll need a solution for this. It's pretty bad if restarting the kubelet means restarting all of the pods on that node.

One option might be to rely on the new CNI CHECK command that was recently introduced to retrieve info for a running pod. However, we'll need to make sure we're still compatible with plugins that implement older versions of the CNI spec for some time because I suspect it will be a while before CHECK implementations become standard.

bowei commented:
cc: @freehan

AlexeyPerevalov commented:
> > Current implementation has functional drawback: if we restart kubelet when pod already started,
> > pod's IP becames none, after that pod will be restarted, and its IP obtained by CNI ADD method
> > is correct. But restarting of the pod is not desirable behaviour.
> 
> I think this is really substantial, and we'll need a solution for this. It's pretty bad if restarting the kubelet means restarting all of the pods on that node.
> 
> One option might be to rely on the new CNI CHECK command that was recently introduced to retrieve info for a running pod. However, we'll need to make sure we're still compatible with plugins that implement older versions of the CNI spec for some time because I suspect it will be a while before CHECK implementations become standard.

One clarification, not all pods are restarting, but only these that doesn't have eth0, because here still nsenter approach is on the stage.
As mentioned @dcbw in previous PR need to get rid of nsenter approach. But in this case pod's ip address should be stored somewhere, it already stored in etcd. So just need to get it from etcd, instead of trying to obtain it from pod at restart time.

dcbw commented:
> As mentioned @dcbw in previous PR need to get rid of nsenter approach. But in this case pod's ip address should be stored somewhere, it already stored in etcd. So just need to get it from etcd, instead of trying to obtain it from pod at restart time

@AlexeyPerevalov latest libcni has caching ability to store the Result that gets passed back to the CRI on-disk somewhere.  It also has a function to grab that back out of the cache, which we anticipated CRI would use on restart when it wanted to get the pod details. Additionally, we are adding a CHECK call that the CRI would call to ensure the container's network was still OK after a restart.

These two things combined should allow kubelet to get the pod network details and verify they are still valid on CRI/kubelet restart.  These changes are anticipated to land "soon" (eg matter of weeks).

AlexeyPerevalov commented:
>  @dcbw

Oh, great ) I already though how to reuse checkpointing in kubernetes for kubelet restart case.

dcbw commented:
@AlexeyPerevalov any chance you want ot resurrect this issue and rebase? Upstream CNI's libcni landed the caching stuff a couple months ago so it should be ready to import into Kube and use.

AlexeyPerevalov commented:
@dcbw I'm going to rebase and check it, during this or next week.

Commit:
Remove nsenter approach and replace it with cni cached result

This commit takes into account ipv6 dual stack feature.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

AlexeyPerevalov commented:
/assign @liggitt

liggitt commented on go.mod:
```diff
@@ -36,7 +36,7 @@ require (
 	github.com/containerd/containerd v1.0.2 // indirect
 	github.com/containerd/typeurl v0.0.0-20190228175220-2a93cfde8c20 // indirect
-	github.com/containernetworking/cni v0.6.0
+	github.com/containernetworking/cni v0.7.0-rc2
```
is there a non-rc tag we can use? v0.7.0?

AlexeyPerevalov replied:
I checked today, 0.7.0 is released. So I'll update patch set soon.

liggitt commented on vendor/github.com/containernetworking/cni/libcni/api.go:
```diff
@@ -15,14 +15,27 @@
 )
 
+var (
+	CacheDir = "/var/lib/cni"
```
does this require this directory be writeable? what happens if the kubelet runs in an environment where this is not writeable?

AlexeyPerevalov replied:
hmm, looks like AddNetwork will return nil instead of result even cni plugin succeeded, with following message:
("failed to set network %q cached result: %v", net.Network.Name, err)
from my point of view it's better to return the result, even access to cache was failed or check cache dir at libcni initialization time.
Looks like after removing nsenter, cni cache became critical for kubelet, especially after kubelet restart. Maybe set CacheDir as a kubelet command line option.

AlexeyPerevalov replied:
it was bad idea to suggest return result in case when cache failed, in this case pod data could be in inconsistent state, we will lost ip address after kubelet restart.

dcbw replied:
@liggitt yes, it requires the directory to be writable. Note that a default install of kubelet using the kubenet plugin also requires a writable /var/lib/cni/networks for host-local already. We could certainly change the default cache dir to be /var/lib/cni/<something> if that's more helpful.

liggitt replied:
I'll defer to sig-node and sig-network to figure out what a guaranteed-writeable dir is

liggitt replied:
a related question is what handles cleanup of stale data in this cache dir? are there changes required to things like `kubeadm reset` to clear this out when resetting a node, or does stale data here get pruned automatically by CNI?

liggitt commented:
/remove-lifecycle rotten

@kubernetes/sig-node-pr-reviews - need feedback on the assumption that `/var/lib/cni` is writeable by the kubelet

@kubernetes/sig-network-pr-reviews this needs review from a network perspective

liggitt commented:
/unassign
/assign @thockin

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -312,40 +316,77 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
+	}
+	// Fall back to first un-associated address for compatibility
+	// with older CNI plugins
+	klog.Warningf("There is no ip address associated with network interface : %s", rt.IfName)
```
@AlexeyPerevalov If the plugin is CNIVersion 0.1.0 or 0.2.0 the Result won't have any Interfaces when upconverted to cnicurrent.Result. Instead of unconditionally printing the warning, I would only print the warning if `len(res.Interfaces) > 0`

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -312,40 +316,77 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 
+func GetIpFromResult(res *cnicurrent.Result, rt *libcni.RuntimeConf) string {
+	klog.V(4).Infof("res: %s", res.String())
+	rt_json, err := json.Marshal(rt)
```
I would make these log messages much more descriptive. And instead of %s (for Result) or marshalling to JSON (for RuntimeConf) why not just log them with %v so the Go object gets printed directly?

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -312,40 +316,77 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 	}
 	klog.V(4).Infof("Added %s to network %s: %v", pdesc, netConf.Name, res)
+
+	if res == nil {
```
You'll never get a nil result if err is also nil. I think we can skip this check.

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -312,40 +316,77 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
+	curRes, err := cnicurrent.NewResultFromResult(res)
+	if curRes != nil && len(curRes.IPs) != 0 {
+		klog.V(4).Infof("IP address for CNI network was set from plugin's output")
+		plugin.SetPodIP(podSandboxID, GetIpFromResult(curRes, rt))
```
Why not just cache the entire Result object, and then extract the IP when we need it? I think we'll have cases to use other fields in result later.

AlexeyPerevalov replied:
Yeah, I agree it can be useful, but name of type Result too general, although it can be normal if it will be just cniOutput.

AlexeyPerevalov replied:
and plugin will keep cni output.

dcbw commented on pkg/kubelet/dockershim/network/cni/cni_test.go:
```diff
@@ -161,35 +167,14 @@ func TestCNIPlugin(t *testing.T) {
 	testBinDir := path.Join(tmpDir, "opt", "cni", "bin")
 	testDataDir := path.Join(tmpDir, "output")
-	defer tearDownPlugin(tmpDir)
+	//defer tearDownPlugin(tmpDir)
```
This seems like a leftover?

AlexeyPerevalov replied:
yes, I removed it to check cni-test results and invoke this test manually.

dcbw commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -61,20 +64,52 @@ func (plugin *cniNetworkPlugin) platformInit() error {
+		klog.Errorf("Error get pod network status when building cni runtime conf: %v", err)
+		return nil, err
+	}
+	res, err := cniNet.GetNetworkCachedResult(netConf, rt)
```
Don't actually need to to through all the buildCNIRuntimeConf() and getting the NS and all that. It's enough to set the rt.ContainerID and rt.IfName (always "eth0" for kubelet) correctly.

And sorry about the "chaining cni?" comment stuff. You don't actually need to do any of this, just call GetNetworkListCachedResult() instead; it was an oversight that wasn't part of the interface but that has been fixed in https://github.com/containernetworking/cni/pull/669 so if you revendor you can grab simplify this code a lot.

AlexeyPerevalov replied:
thanks, can I rely on GetNetworkListCachedResult will be in 0.7.0 libcni?

bboreham replied:
0.7.0 was released before that change; 0.7.1 or later will have it.

dcbw commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -61,20 +64,52 @@ func (plugin *cniNetworkPlugin) platformInit() error {
 	}
+	klog.V(4).Infof("Get cached result for %s in the network %s: %v", pdesc, netConf.Network.Name, res)
+
+	if res == nil {
```
Same here as above, you won't get res = nil if err is also nil.

dcbw commented:
Since both this and another issue need to bump CNI, I pushed PR https://github.com/kubernetes/kubernetes/pull/78908 to do this and address @liggit's directory question.

dcbw commented:
/retest

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -56,13 +61,14 @@ type cniNetworkPlugin struct {
-	podCidr     string
+	host         network.Host
+	execer       utilexec.Interface
+	nsenterPath  string
```
@AlexeyPerevalov nsenterPath is now unused.

AlexeyPerevalov replied:
yes, done

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -318,18 +325,50 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 	return fmt.Sprintf("%s_%s/%s", namespace, name, id.ID)
 }
 
+func GetIpFromResult(res *cnicurrent.Result, ifaceName string) []net.IP {
```
Now should be GetIpsFromResult()?

AlexeyPerevalov replied:
ok, done

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -318,18 +325,50 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 }
 
-func podDesc(namespace, name string, id kubecontainer.ContainerID) string {
+func PodDesc(namespace, name string, id kubecontainer.ContainerID) string {
```
Any particular reason this is now exported (eg initial capital)?

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -362,6 +410,22 @@ func (plugin *cniNetworkPlugin) deleteFromNetwork(network *cniNetwork, podName s
 	return nil
 }
 
+func (plugin *cniNetworkPlugin) SetPodCNIResult(podSandboxID kubecontainer.ContainerID, cniResult *cnicurrent.Result) {
```
This doesn't need to be exported either.

AlexeyPerevalov replied:
agree

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -362,6 +410,22 @@ func (plugin *cniNetworkPlugin) deleteFromNetwork(network *cniNetwork, podName s
+	}
+}
+
+func (plugin *cniNetworkPlugin) GetPodIPs(podSandboxID kubecontainer.ContainerID) []net.IP {
```
Same here, doesn't need to be exported.

dcbw commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -61,19 +63,53 @@ func (plugin *cniNetworkPlugin) platformInit() error {
 	if err != nil {
+		klog.Errorf("Error get cached result %s for network %s/%s: %v", pdesc,
+			netConfList.Plugins[len(netConfList.Plugins) - 1].Network.Type,
+			netConfList.Plugins[len(netConfList.Plugins) - 1].Network.Name, err)
```
@AlexeyPerevalov the result is for the whole network list, not just a single plugin, so it doesn't make sense to log a plugin type here. Also, you can simply do "netConfList.Name" becuase CNI ensures that the network Name in the NetworkConfigList object is the same for all plugins in the list.

dcbw commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -61,19 +63,53 @@ func (plugin *cniNetworkPlugin) platformInit() error {
+	}
+
+	klog.V(4).Infof("Get cached result for %s in the network %s: %v", pdesc,
+		netConfList.Plugins[len(netConfList.Plugins) - 1].Network.Name, res)
```
Same here, just netConfList.Name

dcbw commented:
/retest

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -318,18 +319,50 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 }
 
-func podDesc(namespace, name string, id kubecontainer.ContainerID) string {
+func PodDesc(namespace, name string, id kubecontainer.ContainerID) string {
```
@AlexeyPerevalov I still think this doesn't need to be capitalized/exported since all the users are in the `package cni`.

AlexeyPerevalov replied:
@dcbw, I forgot to ask (pkg/kubelet/util/format/pod.go) has a similar function, podDesc. Maybe reuse it.

dcbw commented on pkg/kubelet/dockershim/network/cni/cni_test.go:
```diff
@@ -74,7 +72,7 @@ func installPluginUnderTest(t *testing.T, testBinDir, testConfDir, testDataDir,
 	f, err = os.Create(pluginExec)
 
-	const execScriptTempl = `#!/usr/bin/env bash
+	execScriptTempl := `#!/usr/bin/env bash
```
@AlexeyPerevalov was this change from `const` necessary? It looks unrelated.

dcbw commented:
@AlexeyPerevalov Just a couple nits but overall looks good and I'm OK once those are fixed and CI passes. Thanks!

dcbw commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -37,6 +39,9 @@ import (
 	utilexec "k8s.io/utils/exec"
+
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
```
@AlexeyPerevalov CI is complaining about these:

```
GoCompile: missing strict dependencies:
	/b/f/w/pkg/kubelet/dockershim/network/cni/cni.go: import of "k8s.io/apiserver/pkg/util/feature"
	/b/f/w/pkg/kubelet/dockershim/network/cni/cni.go: import of "k8s.io/kubernetes/pkg/features"
Known dependencies are:
```

perhaps they aren't in the Bazel files?

warmchang commented:
/retest

AlexeyPerevalov commented:
/retest

cmluciano commented:
/ok-to-test

AlexeyPerevalov commented:
/test pull-kubernetes-node-e2e

AlexeyPerevalov commented:
/retest

aojea commented:
/cc @khenidak

AlexeyPerevalov commented:
/test pull-kubernetes-e2e-gce-100-performance

AlexeyPerevalov commented:
/test pull-kubernetes-kubemark-e2e-gce-big

johscheuer commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -331,8 +334,36 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
+		klog.V(3).Info("res is nil")
+		return nil
+	}
+	ipv6DualStack := utilfeature.DefaultFeatureGate.Enabled(kubefeatures.IPv6DualStack)
```
IMHO the check if `IPv6DualStack` is enabled is not required. We could  remove the `if ipv6DualStack` check and just return always the `outlist` as a result (for non `IPv6DualStack` cluster with one item otherwise with two).

AlexeyPerevalov replied:
done

AlexeyPerevalov commented:
/test pull-kubernetes-integration

AlexeyPerevalov commented:
/test pull-kubernetes-integration

AlexeyPerevalov commented:
/test pull-kubernetes-kubemark-e2e-gce-big

johscheuer commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -345,8 +345,30 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 
-func podDesc(namespace, name string, id kubecontainer.ContainerID) string {
-	return fmt.Sprintf("%s_%s/%s", namespace, name, id.ID)
+func getIpsFromResult(res *cnicurrent.Result, ifaceName string) []net.IP {
```
Any reason this method isn't exported ? This could be useful for kubenet too: https://github.com/kubernetes/kubernetes/pull/82844/files#r329884258

johscheuer replied:
I can change this in my PR no need to do this in here. Just a question if there are any reasons not to use this method in `kubenet`.

AlexeyPerevalov replied:
@johscheuer, @dcbw I think if function is used by libraries which are on the same level (layer), like cni/kubenet we need to move that function to common library. Is k8s.io/kubernetes/pkg/kubelet/dockershim/network a common library for this case?

johscheuer replied:
IMHO I would add here a second method: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/network/plugins.go#L260 e.g.`GetPodIPsFromCNIResult`

AlexeyPerevalov replied:
Ok, I can split this commit and add this method there, but tomorrow.

Commit:
Introduce helper GetIpsFromResult

This function returns ip addresses from cni result.
It could be used in cni and kubenet packages.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

johscheuer commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-100-performance
/test pull-kubernetes-integration

johscheuer commented:
/lgtm @dcbw could you also take a look ?

AlexeyPerevalov commented:
Now I think that containernetworking library is also proper place for GetIpsFromResult

neolit123 commented:
/remove-area kubeadm

AlexeyPerevalov commented:
/test pull-kubernetes-e2e-gce-storage-slow

AlexeyPerevalov commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

AlexeyPerevalov commented:
/test pull-kubernetes-kubemark-e2e-gce-big

AlexeyPerevalov commented:
/test pull-kubernetes-e2e-gce-100-performance

danwinship commented:
ping @dcbw

AlexeyPerevalov commented:
/test pull-kubernetes-verify

Vladimare commented:
/retest

AlexeyPerevalov commented:
/retest

AlexeyPerevalov commented:
/test pull-kubernetes-verify

AlexeyPerevalov commented:
ping @dcbw

bboreham reviewed (commented):
Thanks!  Overall seems good, but I was concerned about one corner case.

AlexeyPerevalov commented on pkg/kubelet/dockershim/network/plugins.go:
```diff
@@ -290,6 +291,32 @@ func GetPodIPs(execer utilexec.Interface, nsenterPath, netnsPath, interfaceName
+	if len(res.Interfaces) > 0 {
+		klog.Warningf("There is no ip address associated with network interface : %s", ifaceName)
+	}
+	return []net.IP{res.IPs[0].Address.IP}
```
Good point! I relied on two places:

1. when saving cniResult in addToNetwork
`if curRes != nil && len(curRes.IPs) != 0 {
                klog.V(4).Infof("IP address for CNI network was set from plugin's output")
                plugin.setPodCNIResult(podSandboxID, curRes)`
 
2.  before calling GetIpsFromResult(curRes,... in GetPodNetworkStatus
        `curRes, err := cnicurrent.NewResultFromResult(res)
        if curRes == nil || len(curRes.IPs) == 0 {
                klog.Errorf("CNI result conversion failed: %v", err)
                return nil, err
        }`

This function can be used somewhere else in future, it's better to make it more robust.

AlexeyPerevalov replied:
done in this func
if res == nil || len(res.IPs) == 0

AlexeyPerevalov commented:
/test pull-kubernetes-e2e-gce-100-performance

AlexeyPerevalov commented:
/test pull-kubernetes-integration

bboreham commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -49,35 +53,55 @@ func getLoNetwork(binDirs []string) *cniNetwork {
+
+	ips := network.GetIpsFromResult(curRes, rt.IfName)
 	if len(ips) == 0 {
 		return nil, fmt.Errorf("cannot find pod IPs in the network namespace, skipping pod network status for container %q", id)
```
This error message needs updating to reflect possible failures in the new code.

bboreham commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -49,35 +53,55 @@ func getLoNetwork(binDirs []string) *cniNetwork {
+		klog.V(3).Infof("get pod ip %v from plugin", podIPs)
+		return &network.PodNetworkStatus{IP: podIPs[0], IPs: podIPs}, nil
+	}
+	cninetwork := plugin.getDefaultNetwork()
```
Everything in this function from here on is concerned with fetching a result from CNI's cache; is that right?

I would probably comment that or create a new function with a name to indicate that.

AlexeyPerevalov replied:
> Everything in this function from here on is concerned with fetching a result from CNI's cache; is that right?
> 
> I would probably comment that or create a new function with a name to indicate that.
yes, you're right. I agree it worth marking it.

bboreham commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -49,35 +53,55 @@ func getLoNetwork(binDirs []string) *cniNetwork {
+		return nil, nil
+	}
 	if err != nil {
+		klog.Errorf("Error get cached result %s for network %s: %v", pdesc, netConfList.Name, err)
```
How can we get here?  What are the expected cases where `res != nil` and `err != nil` ?

AlexeyPerevalov replied:
probably I just wanted to return err if it existed, yes, it worth merging both condition.

AlexeyPerevalov commented on pkg/kubelet/dockershim/network/cni/cni.go:
```diff
@@ -351,18 +347,14 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 	}
 
-	pdesc := podDesc(podNamespace, podName, podSandboxID)
+	pdesc := format.PodDesc(podName, podNamespace, machinerytypes.UID(podSandboxID.ID))
```
I moved it into separate PR (https://github.com/kubernetes/kubernetes/pull/94251), since it's just code cleanup.

AlexeyPerevalov commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -67,23 +61,47 @@ func (plugin *cniNetworkPlugin) GetPodNetworkStatus(namespace string, name strin
+	}
+	res, err := cniNet.GetNetworkListCachedResult(netConfList, rt)
 
-	ips, err := network.GetPodIPs(plugin.execer, plugin.nsenterPath, netnsPath, network.DefaultInterfaceName)
```
I keep kubenet_linux untouched, and network.GetPodIPs still used there. IOW kubenet still goes into network namespace and checks for eth0.

AlexeyPerevalov commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -49,35 +53,55 @@ func getLoNetwork(binDirs []string) *cniNetwork {
-}
-
 // TODO: Use the addToNetwork function to obtain the IP of the Pod. That will assume idempotent ADD call to the plugin.
 // Also fix the runtime's call to Status function to be done only in the case that the IP is lost, no need to do periodic calls
```
last part, line 57 is not appropriate

AlexeyPerevalov commented on pkg/kubelet/dockershim/network/cni/cni_others.go:
```diff
@@ -49,35 +53,55 @@ func getLoNetwork(binDirs []string) *cniNetwork {
-	return nil
-}
-
 // TODO: Use the addToNetwork function to obtain the IP of the Pod. That will assume idempotent ADD call to the plugin.
```
It's not but it's better to actualize in this PR or even in another, since it's no more necessary to call addNetwork.

neolit123 commented:
/remove-sig cluster-lifecycle
